torch/_higher_order_ops/triton_kernel_wrap.py

```diff
@@ -60,88 +62,235 @@ def reset_table(self) -> None:
-                self.read_depth -= 1
-                return
-        self.generic_visit(node)
-
-
-def filter_non_mutated(kernel, tensors):
+            nonlocal next_fake_intermediate
+            next_fake_intermediate -= 1
+            return Intermediate(next_fake_intermediate)
+
+        assert isinstance(token, lark.lexer.Token)
+
+        if token.type == "INTERMEDIATE":
+            return Intermediate(int(token.value[len("%") :]))
+        if token.type == "PARAM":
+            return Param(int(token.value[len("%arg") :]))
+
+        raise AssertionError(f"{type(token.type)} => {token.value} invalid")
+
+    # In alternative representation, function names are quoted.
+    # It should be possible to move this into the grammar alltogether.
+    def convert_name(token):
+        s = token.value
+        if s[0] == '"' and s[-1] == '"':
+            return s[1:-1]
+        return s
+
+    @v_args(inline=True)
+    class CalculateOps(Transformer):
+        def process_op_no_ret(self, *args):
+            self.process_op(None, *args)
+
+        def process_op(self, ret, name, args, *rest):
+            ops[convert(ret)] = Op(convert_name(name), convert(args))
+
+    parser = Lark(grammar, parser="lalr", transformer=CalculateOps())
+    parser.parse(ttir)
+    return ops
+
+
+# Given a triton kernel and the arguments for this kernel, this function
+# retrieves the TTIR converted version of the kernel from Triton's API.
+# Parsing the TTIR, it creates a control flow graph, detects all sinks
+# from a predefined list of sinks. NOTE: What if triton exposed this MemWrite Trait
+# From each sink, it traverses the CFG backwards to identify all the input
+# pointers that are mutated
+def identify_mutated_tensors(kernel, kwargs):
+    from triton.compiler.code_generator import ast_to_ttir
     from triton.runtime.autotuner import Autotuner
+    from triton.runtime.jit import JITFunction
+
+    import torch
+    from torch._subclasses.fake_tensor import FakeTensor
 
     if isinstance(kernel, Autotuner):
+        if len(kernel.configs) > 0:
+            # If we are autotuning, then it doesn't matter which version gets
+            # picked for tracing purposes, so lets pick the first one
+            kwargs = {**kwargs, **kernel.configs[0].kwargs}
         kernel = kernel.fn
 
-    infos = {name: MutationInfo() for name in tensors}
-    tracker = MutationTracker(infos)
-    tracker.visit(kernel.parse())
+    assert isinstance(kernel, JITFunction)
+
+    # Drop the keys
+    # Replace all SymExprs with a regular value for TTIR generation
+    # Replace all FakeTensor with real tensors
+    # These replacements are needed for triton's type, key and config functions
+    args: List[Any] = []
+    for a in kwargs.values():
+        if isinstance(a, (torch.SymInt, torch.SymFloat, torch.SymBool)):
+            args.append(2)
+        elif isinstance(a, FakeTensor):
+            args.append(torch.empty(2, dtype=a.dtype))
```

Comment:
You should check if inputs alias each other

Reply:
This is a problem irrelevant of this PR, current functionalization does not handle cloning and then reassigning the clones back. I want to plan aliased inputs to the kernel all together, there's a TODO in code about this already.

Reply:
> I want to plan aliased inputs to the kernel all together, there's a TODO in code about this already.

In internal kernels, we do have multiple inputs pointing to the same memory region (slices thereof), but all those are inputs / read-only. So, not sure what "plan" means above, but I wouldn't ban this case.

Reply:
We can be more creative and say that only read only inputs can be aliases. Currently if a kernel takes a and a.view(-1) as arguments, after functionalization, they will point to different things. So this is a preexisting bug.